### PR TITLE
ER-509 - Logging update & retry policy

### PR DESCRIPTION
**Pull request by leewadhams**
- Updated logging to resolve issues highlighted in Gorilla testing session.
- Added retry logic to repo/store classes in the event of a MongoCommandException. 

There is a knock-on effect of the retry logic for the jobs, in that if they pick messages off of queues they could retry up to 15 times.

5 (times message retried) **x** 3 (Polly retries)

### CODE REVIEW

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyRepository.cs [hunk 1]
```diff
@@ -20,20 +21,20 @@ internal sealed class MongoDbVacancyRepository : MongoDbCollectionBase, IVacancy
...
         {
             var collection = GetCollection<Vacancy>();
-            await collection.InsertOneAsync(vacancy);
+            await RetryPolicy.ExecuteAsync(() => collection.InsertOneAsync(vacancy));
         }
 
         public async Task<Vacancy> GetVacancyAsync(long vacancyReference)
         {
-            var vacancy = await FindVacancy(v => v.VacancyReference, vacancyReference);
+            var vacancy = await RetryPolicy.ExecuteAsync(() => FindVacancy(v => v.VacancyReference, vacancyReference));
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyRepository.cs]: FindVacancy is already retrying

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyRepository.cs [hunk 2]
```diff
@@ -43,7 +44,7 @@ public MongoDbVacancyRepository(IOptions<MongoDbConnectionDetails> details)
 
         public async Task<Vacancy> GetVacancyAsync(Guid id)
         {
-            var vacancy =  await FindVacancy(v => v.Id, id);
+            var vacancy = await RetryPolicy.ExecuteAsync(() => FindVacancy(v => v.Id, id));
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyRepository.cs]: FindVacancy is already retrying

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Mongo/MongoDbCollectionBase.cs [hunk 3]
```diff
@@ -28,5 +38,33 @@ protected IMongoCollection<T> GetCollection<T>()
...
+        }
+
+        protected Task<T> Retry<T>(Func<Task<T>> func)
+        {
+            var policy = GetRetryPolicy();
+
+            return policy.ExecuteAsync(func);
+        }
+
+        private Polly.Retry.RetryPolicy GetRetryPolicy()
```
Comment (chrisjensenuk) [inline src/Shared/Recruit.Vacancies.Client/Infrastructure/Mongo/MongoDbCollectionBase.cs]: Should we inject the policy in?  Having a back-off policy is OK for Jobs but for the web we might want to fail quicker.  Maybe we should revisit this after we've released.

File: src/Jobs/Recruit.Vacancies.Jobs/DashboardGenerator/DashboardGeneratorJob.cs [hunk 4]
```diff
@@ -8,7 +8,7 @@
 
 namespace Esfa.Recruit.Vacancies.Jobs.DashboardGenerator
 {
-    public class DashboardGeneratorJob
+    public partial class DashboardGeneratorJob
```
[the remark this answers is not in the document]
Comment (leewadhams) [reply src/Jobs/Recruit.Vacancies.Jobs/DashboardGenerator/DashboardGeneratorJob.cs]: 👍  Nope it shouldn't.

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyRepository.cs (hunk 1 above)
Comment (leewadhams) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/Repositories/MongoDbVacancyRepository.cs]: 👍 Got carried away :-)

File: src/Shared/Recruit.Vacancies.Client/Infrastructure/Mongo/MongoDbCollectionBase.cs (hunk 3 above)
Comment (leewadhams) [reply src/Shared/Recruit.Vacancies.Client/Infrastructure/Mongo/MongoDbCollectionBase.cs]: Yes I agree we should. Have added an item on tech debt list.

**Pull request merged**